packages/playground-ui/src/components/assistant-ui/attachments/attachment.tsx

```diff
@@ -0,0 +1,124 @@
+'use client';
+
+import { AttachmentPrimitive, AttachmentState, ComposerPrimitive, useAttachment } from '@assistant-ui/react';
+import { TooltipProvider } from '@radix-ui/react-tooltip';
+import { CircleXIcon, PaperclipIcon } from 'lucide-react';
+import { useEffect, useState } from 'react';
+
+import { TooltipIconButton } from '@/components/assistant-ui/tooltip-icon-button';
+import { Tooltip, TooltipContent, TooltipTrigger } from '@/components/ui/tooltip';
+
+import { Icon } from '@/ds/icons';
+import { useHasAttachments } from '../hooks/use-has-attachments';
+import { useAttachmentSrc } from '../hooks/use-attachment-src';
+import { ImageEntry, TxtEntry, PdfEntry } from './attachment-preview-dialog';
+import Spinner from '@/components/ui/spinner';
+import { useLoadBrowserFile } from '../hooks/use-load-browser-file';
+import { fileToBase64 } from '@/lib/file';
+
+const ComposerTxtAttachment = ({ document }: { document: AttachmentState }) => {
+  const { isLoading, text } = useLoadBrowserFile(document.file);
+
+  return (
+    <div className="flex items-center justify-center h-full w-full">
+      {isLoading ? <Spinner className="animate-spin" /> : <TxtEntry data={text} />}
+    </div>
+  );
+};
+
+const ComposerPdfAttachment = ({ document }: { document: AttachmentState }) => {
+  const [state, setState] = useState({ isLoading: false, text: '' });
+  useEffect(() => {
+    const run = async () => {
+      if (!document.file) return;
+      setState(s => ({ ...s, isLoading: true }));
+      const text = await fileToBase64(document.file);
+      setState(s => ({ ...s, isLoading: false, text }));
+    };
+    run();
```

Comment:
make sure component is not unmounted before setting state
```suggestion
    let isCancelled = false;
    const run = async () => {
      if (!document.file) return;
      setState(s => ({ ...s, isLoading: true }));
      const text = await fileToBase64(document.file);
      if (isCanceled) { return; }
      setState(s => ({ ...s, isLoading: false, text }));
    };
    run();
    
    return {
      isCanceled = true;
    }
```